cli/src/commands/asset.ts

```diff
@@ -97,36 +97,60 @@ export const checkForDuplicates = async (files: string[], { concurrency, skipHas
 
   progressBar.start(files.length, 0);
 
-  const newFiles: string[] = [];
-  const duplicates: Asset[] = [];
+  const results: { id: string; checksum: string }[] = [];
 
-  const queue = new Queue<string[], AssetBulkUploadCheckResults>(
-    async (filepaths: string[]) => {
-      const dto = await Promise.all(
-        filepaths.map(async (filepath) => ({ id: filepath, checksum: await sha1(filepath) })),
-      );
-      const response = await checkBulkUpload({ assetBulkUploadCheckDto: { assets: dto } });
-      const results = response.results as AssetBulkUploadCheckResults;
-      for (const { id: filepath, assetId, action } of results) {
-        if (action === Action.Accept) {
-          newFiles.push(filepath);
-        } else {
-          // rejects are always duplicates
-          duplicates.push({ id: assetId as string, filepath });
-        }
-      }
-      progressBar.increment(filepaths.length);
+  const queue = new Queue<string, AssetBulkUploadCheckResults>(
+    async (filepath: string): Promise<AssetBulkUploadCheckResults> => {
+      const dto = { id: filepath, checksum: await sha1(filepath) };
+
+      results.push(dto);
+
+      progressBar.increment();
       return results;
     },
     { concurrency, retry: 3 },
   );
 
-  for (const items of chunk(files, concurrency)) {
-    await queue.push(items);
-  }
+  await Promise.all(
+    files.map(async (item) => {
+      await queue.push(item);
+    }),
+  );
 
   await queue.drained();
 
+  const newFiles: string[] = [];
+  const duplicates: Asset[] = [];
+
+  let retries = 0;
+  const maxRetries = 3;
+
+  while (retries < maxRetries) {
+    try {
+      const chunks = chunk(results, 1000);
+
+      for (const chunk of chunks) {
```

Comment:
It'd be cool to process these chunks while the queue is active so it can do the hashing and requests in parallel. After `results.push(dto);`, you could check if the list is 1000 elements and either do the `checkBulkUpload` in the same job, or add it to a separate queue. In the former, you'd have to be a little more careful about the retries since you don't want an error in `checkBulkUpload` to cause it to add the same asset to `results` twice.